XS.pm

```diff
@@ -1858,17 +1860,23 @@ the encoding for you (for example, when printing to a terminal using a
 filehandle that transparently encodes to UTF-8 you certainly do NOT want
 to UTF-8 encode your data first and have Perl encode it another time).
 
-=item C<utf8> flag enabled
+=item C<utf8> is UTF8_STANDARD
 
-If the C<utf8>-flag is enabled, C<encode>/C<decode> will encode all
+If C<utf8> is UTF8_STANDARD, C<encode>/C<decode> will encode all
 characters using the corresponding UTF-8 multi-byte sequence, and will
 expect your input strings to be encoded as UTF-8, that is, no "character"
 of the input string must have any value > 255, as UTF-8 does not allow
 that.
 
-The C<utf8> flag therefore switches between two modes: disabled means you
-will get a Unicode string in Perl, enabled means you get an UTF-8 encoded
-octet/binary string in Perl.
+=item C<utf8> is UTF8_BYTES
+
+When C<utf8> is UTF8_BYTES, C<encode> and C<decode> generate and expect
+UTF-8 strings. This ends up being I<almost> the same as UTF_CHARS except
+that JSON Unicode escapes (e.g., C<"\u00e9">) are output encoded as UTF-8.
+
+This mode also allows arbitrary octet sequences to be decoded. Ordinarily
+that’s undesirable (since it violates the JSON specification), but in some
+applications it may nonetheless be appropriate.
```

Comment:
This sort of documentation is a good way to end up with thousands of buggy programs that should not have used it.

Reply:
This doesn’t seem to have impeded the addition of binary() mode. The documentation’s suggestions of UTF-16 and UTF-32 remain, despite that JSON (now) forbids these. The original latin1() mode remains, too.

All of these things “should not” happen because they violate the JSON spec.

Reply:
These are all internally consistent operations that used to be allowed by the spec though they are rather pointless in new code, there's plenty of history to this module (and particularly JSON::PP which it is compatible with). More importantly, they won't cause bugs if used with data of the appropriate encoding.

Reply:
> More importantly, they won't cause bugs if used with data of the appropriate encoding.

That’s the point of this proposal as well: to facilitate proper end-to-end functionality, given the need to allow arbitrary octet sequences.

You asserted earlier:
> there is no way to preserve arbitrary octets while also correctly interpreting unicode escapes except by interpreting the octets as characters instead of bytes.

I _believe_ this PR disproves your statement. I’ve added a test to verify it explicitly.

Can you show a counterexample?

Reply:
This functionality cannot accomplish it by definition, because the unicode escapes are not decoded to the specified unicode characters.

Reply:
Ohh-kay, by “correct” you specifically mean that `"\u00e9"` becomes Perl `\xe9`.

I’ve got nothing for you that I haven’t written previously. We need the Unicode escapes to end up as UTF-8, and we can’t get there by encoding Unicode. If @rurban wants to allow that, I’ll polish this up. If not, we’ll patch locally.